### PR TITLE
Correctness of Add

### PR DESCRIPTION
**Pull request by ChrisPattison**
This adds
- Much more aggressive testing of inputs. Generated inputs have random signs, contain 0.0 and 1.0, and include random exponents according to a Poisson distribution. Under this testing a large portion of the test cases failed originally
- PackedFloat <-> MPFR conversions are heavily modified to handle edge cases
- UnitTest comparisons are slightly modified because the conversion through PackedFloat throws away the mantissa when the value is singular (ex. 0, NaN, inf, etc)
- Add is heavily modified to avoid subnormal outputs and correctly handle different sign bits on the input.

Note: The intent with Add right now is for correctness, so I have commented out some of the techniques required for synthesis like the barrel shifter. Currently, Add fails the unit tests in two ways: the least significant bit of the most significant limb is sometimes wrong, and sometimes the output is subnormal

### CODE REVIEW

File: include/PackedFloat.h [hunk 1]
```diff
@@ -149,6 +172,19 @@ class PackedFloat {
         return !(*this == rhs);
     }
 
+    /// Slightly looser comparison where zero values are allowed to differ in the sign and exponent
+    inline bool EquivCompare(PackedFloat const &rhs) {
```
Comment (definelicht) [inline include/PackedFloat.h]: I would be in favor of adding this to the regular equals operator. Is there a case where we want the stricter compare?

File: include/Random.h [hunk 2]
```diff
@@ -32,6 +33,13 @@ class RandomNumberGenerator {
     void Generate(mpfr_ptr);
 
    private:
+    std::mt19937_64 small_rng_;
+    static constexpr double neg_frac_ = 1.0/3.0;
```
Comment (definelicht) [inline include/Random.h]: The naming convention for compile-time constants (from Google's style guide) is `kMyConstant`

File: device/ArithmeticOperations.cpp [hunk 3]
```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
...
+    const Exponent shift_c = (!a_exp_is_larger && !b_mant_is_zero) ? (b.exponent - a.exponent) : 0;
+    const Exponent shift_m = (a_exp_is_larger && !a_mant_is_zero) ? (a.exponent - b.exponent) : 0;
+
+#ifndef HLSLIB_SYNTHESIS
+    // Turns out Xilinx allows signed shifts!
+    assert(shift_m >= 0);
+    assert(shift_c >= 0);
+#endif
+
+    // Optionally shift right by 1, 2, 4, 8, 16... log2(B), such that all bits have eventually traveled to
```
Comment (definelicht) [inline device/ArithmeticOperations.cpp]: Move down to the commented out code that it refers to

File: device/ArithmeticOperations.cpp [hunk 4]
```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
...
-    const int kNumStages = hlslib::ConstLog2(kBits);
-ShiftStages:
-    for (int i = 0; i < kNumStages; ++i) {
-#pragma HLS UNROLL
-        a_mantissa = ((shift_c & (1 << i)) == 0) ? a_mantissa : (a_mantissa >> (1 << i));
-        b_mantissa = ((shift_m & (1 << i)) == 0) ? b_mantissa : (b_mantissa >> (1 << i));
-    }
+    // This is unsigned
+    a_mantissa = a_mantissa >> shift_c;
+    b_mantissa = b_mantissa >> shift_m;
```
Comment (definelicht) [inline device/ArithmeticOperations.cpp]: This is definitely nicer (the old code was a remnant from when we used a language that didn't allow dynamic shifts). Did you compare the reports or the generated RTL between the two, to make sure it does something reasonable?

File: device/ArithmeticOperations.cpp [hunk 5]
```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
...
-    return {m_sign, m_exponent, &m_mantissa};
+    PackedFloat result;
+    std::memcpy(result.mantissa, &m_mantissa, kMantissaBytes);
+    result.exponent = m_exponent;
+    result.sign = a.sign != b.sign;
+    return result;
 }
 
+// Does this correctly output the result if a and b are different signs?
+// The mantissa of the result should depend on the sign bits of a and b
```
Comment (definelicht) [inline device/ArithmeticOperations.cpp:56]: No, we only support positive, non-subnormal, non-special value numbers, since this should be all we need for the matmul.

File: device/ArithmeticOperations.cpp [hunk 6]
```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
...
+// ShiftStages:
+//     for (int i = 0; i < kNumStages; ++i) {
+// #pragma HLS UNROLL
+//         a_mantissa = ((shift_c & (1 << i)) == 0) ? a_mantissa : (a_mantissa >> (1 << i));
+//         b_mantissa = ((shift_m & (1 << i)) == 0) ? b_mantissa : (b_mantissa >> (1 << i));
+//     }
     // Finally zero out the mantissas if they are shifted by more than the precision
-    a_mantissa = shift_c >= kBits ? decltype(a_mantissa)(0) : a_mantissa;
-    b_mantissa = shift_m >= kBits ? decltype(b_mantissa)(0) : b_mantissa;
+    // Is this necessary?
```
Comment (definelicht) [inline device/ArithmeticOperations.cpp]: It was necessary with the old method of shifting, since the unrolled loop only implements "useful" shifts. I think you can comment it out along with the old shifter, assuming it still passes tests

**Review by definelicht — CHANGES_REQUESTED** [review]
Thanks, but why spend time on this now? We just need it to work for the matrix multiplies we would see in SDPB -- the edge cases don't matter for now.

The important is that we don't do worse on our use case than before. The bugs you are still seeing in validation, are they new, or did they exist before?

File: include/PackedFloat.h (hunk 1 above)
Comment (ChrisPattison) [reply include/PackedFloat.h]: It seemed like you were going for exact bit compatibility with MPFR. This comparison relaxes that slightly for 0

Comment (definelicht) [reply include/PackedFloat.h]: I would like to follow MPFR semantics as closely as possible, so if this is how they treat zero, and it's no significant extra work for us to do the same, I think we should too

File: device/ArithmeticOperations.cpp (hunk 4 above)
Comment (ChrisPattison) [reply device/ArithmeticOperations.cpp]: I have not. I was just eliminating code for the moment. If you think this has a reasonable shot at generating sane hardware then I'll give it a shot

File: include/PackedFloat.h (hunk 1 above)
Comment (ChrisPattison) [reply include/PackedFloat.h]: I think it's a fair bit of work to get exact bit compatibility of the mantissa at zero with MPFR. What would be reasonable is encoding the singular values + zero in the exponent like MPFR does. For context, currently we are encoding that zero mantissa = zero.

Here's the relevant definition from the MPFR header
Image Pasted at 2022-1-3 22-19
o

File: device/ArithmeticOperations.cpp (hunk 4 above)
Comment (definelicht) [reply device/ArithmeticOperations.cpp]: No clue :-) It might, and it might be bother worse and better than our homemade version, so we have to check

File: include/PackedFloat.h (hunk 1 above)
Comment (ChrisPattison) [reply include/PackedFloat.h]: Per our convo, we'll circle back to this later
